gk/fib.c

```diff
@@ -847,15 +939,16 @@ ether_cache_put(struct gk_fib *neigh_fib,
 	enum gk_fib_action action, struct ether_cache *eth_cache,
 	struct ipaddr *addr, struct gk_config *gk_conf)
 {
+	int ret;
 	struct gk_fib *neighbor_fib = neigh_fib;
 
-	if (eth_cache->ref_cnt == 0) {
+	if (rte_atomic32_read(&eth_cache->ref_cnt) == 0) {
 		rte_panic("Unexpected condition: the ref_cnt of the ether cache is 0 at %s\n",
 			__func__);
 	}
 
-	if (eth_cache->ref_cnt > 1) {
-		eth_cache->ref_cnt--;
+	if (rte_atomic32_read(&eth_cache->ref_cnt) > 1) {
+		rte_atomic32_dec(&eth_cache->ref_cnt);
```

Comment:
Notice that we are dealing with atomic operations here, so we cannot read and then update because the value could've changed since the time we read it. There's a function, which I don't remember the name, that tests and, if possible, decrements atomically. The new function will even allow you to fold the previous `if` statement into it since one cannot decrement from zero.

Reply:
After looking up the atomic functions in DPDK, I found that the definition of the atomic value actually is ``int32_t`` type, so the counter could potentially be negative. I found two functions: ``rte_atomic32_sub_return()`` atomically subtracts a 32-bit value from a counter and return the result; ``rte_atomic32_dec_and_test()`` atomically decrements the atomic counter by one and returns true if the result is 0, or false in all other cases. I think the former one may be a better fit.

Reply:
Agree. `rte_atomic32_sub_return()` is good enough for our needs.